### PR TITLE
Allow delegates to delete themselves

### PR DESCRIPTION
**Pull request by Uxio0**

### CODE REVIEW

File: safe_transaction_service/history/serializers.py [hunk 1]
```diff
@@ -258,15 +259,33 @@ class SafeDelegateDeleteSerializer(serializers.Serializer):
...
+
+    def get_valid_delegators(self, ethereum_client: EthereumClient, safe_address: ChecksumAddress,
+                             delegate: ChecksumAddress) -> List[ChecksumAddress]:
+        """
+        :param ethereum_client:
+        :param safe_address:
+        :param delegate:
+        :return: Valid delegators for a Safe. A delegate should be able to remove itself
+        """
+        return self.get_safe_owners(ethereum_client, safe_address) + [delegate]
```
Comment (jpalvarezl) [inline safe_transaction_service/history/serializers.py:277]: Are owners considered `delegates` in this method?

File: safe_transaction_service/history/serializers.py [hunk 2]
```diff
@@ -321,6 +334,16 @@ def validate(self, data):
...
 
+    def get_valid_delegators(self, ethereum_client: EthereumClient, safe_address: ChecksumAddress,
+                             delegate: ChecksumAddress) -> List[ChecksumAddress]:
+        """
+        :param ethereum_client:
+        :param safe_address:
+        :param delegate:
+        :return: Valid delegators for a Safe. A delegate shouldn't be able to add itself
+        """
+        return self.get_safe_owners(ethereum_client, safe_address)
```
Comment (jpalvarezl) [inline safe_transaction_service/history/serializers.py:345]: This is a bit confusion as this appears to be returning owners when the method should return delegators 🤔

File: safe_transaction_service/history/tests/test_views.py [hunk 3]
```diff
@@ -1091,6 +1091,17 @@ def test_delete_safe_delegate(self):
         self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
         self.assertEqual(SafeContractDelegate.objects.count(), 1)
 
+        # Check if a delegate can delete itself
+        SafeContractDelegate.objects.all().delete()
+        delegate_account = Account().create()
+        SafeContractDelegateFactory(safe_contract=safe_contract, delegate=delegate_account.address)
+        hash_to_sign = DelegateSignatureHelper.calculate_hash(delegate_account.address)
+        data['signature'] = delegate_account.signHash(hash_to_sign)['signature'].hex()
+        response = self.client.delete(reverse('v1:history:safe-delegate', args=(safe_address, delegate_account.address)),
```
Comment (jpalvarezl) [inline safe_transaction_service/history/tests/test_views.py:1100]: I am not sure if this already behaves like so, but if we allow to delete several delagators, should we make this a transactional method?

**Review by jpalvarezl — APPROVED** [review]
I have a couple of questions, regarding mostly what the difference of an `owner` and a `delegator` is in this PR. I am just a bit confusded by the methods used in some places. I would approve just not to block you.

File: safe_transaction_service/history/serializers.py (hunk 1 above)
Comment (Uxio0) [reply safe_transaction_service/history/serializers.py:277]: No, owners are delegators

File: safe_transaction_service/history/serializers.py (hunk 2 above)
Comment (Uxio0) [reply safe_transaction_service/history/serializers.py:345]: Owners are the delegators (they delegate in a `delegate`)

File: safe_transaction_service/history/tests/test_views.py (hunk 3 above)
Comment (Uxio0) [reply safe_transaction_service/history/tests/test_views.py:1100]: What do you mean?

**Pull request merged**